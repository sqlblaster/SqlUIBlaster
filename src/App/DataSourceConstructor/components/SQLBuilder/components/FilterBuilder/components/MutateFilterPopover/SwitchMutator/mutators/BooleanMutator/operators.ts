enum BooleanOperatorEnum {
  'Is true',
  'Is false',
  'Is empty',
  'Is not empty'
}

export type BooleanOperator = keyof typeof BooleanOperatorEnum;

export const booleanOperators: BooleanOperator[] = [
  'Is true',
  'Is false',
  'Is empty',
  'Is not empty'
];
