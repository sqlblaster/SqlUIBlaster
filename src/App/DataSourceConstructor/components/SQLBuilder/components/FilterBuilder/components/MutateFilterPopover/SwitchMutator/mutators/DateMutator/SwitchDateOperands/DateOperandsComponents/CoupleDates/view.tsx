import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { DatePickerByOperandsType } from '../../DatePickerByOperandsType';
import { CoupleDateViewProps } from './props';
import { coupleDateStyles } from './styles';

export const CoupleDate = withStyles(coupleDateStyles)((({
  operatorView,
  operands,
  pickedVariables,
  classes,
  handleVariablePick,
  handleOperandsChange
}) => {
  return (
    <>
      <div className={classes.root}>
        {operatorView}
        <div className={classes.dates}>
          <div className={classes.date} data-testid='between-operand-1'>
            <DatePickerByOperandsType
              operands={operands[0]}
              onOperandsChange={handleOperandsChange(0)}
              variable={pickedVariables[0]}
              onVariablePicked={handleVariablePick(0)}
              limits={
                operands[1].dateType === 'Fixed date'
                  ? {
                      maxDate: operands[1].date,
                      maxDateMessage:
                        'Couldn\'t set time, because selected date is later than second date'
                    }
                  : undefined
              }
            />
          </div>
          <div className={classes.date} data-testid='between-operand-2'>
            <DatePickerByOperandsType
              operands={operands[1]}
              onOperandsChange={handleOperandsChange(1)}
              variable={pickedVariables[1]}
              onVariablePicked={handleVariablePick(1)}
              limits={
                operands[0].dateType === 'Fixed date'
                  ? {
                      minDate: operands[0].date,
                      minDateMessage:
                        'Couldn\'t set time, because selected date is earlier than first date'
                    }
                  : undefined
              }
            />
          </div>
        </div>
      </div>
    </>
  );
}) as React.FC<CoupleDateViewProps>);
