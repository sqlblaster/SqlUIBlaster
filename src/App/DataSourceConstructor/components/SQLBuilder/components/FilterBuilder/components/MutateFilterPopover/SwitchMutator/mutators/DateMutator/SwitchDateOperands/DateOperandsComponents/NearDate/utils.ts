import { PluralDateComponent } from './plural-date-components';

export const getThisDateComponentNotation = (
  dateComponent: PluralDateComponent
) => {
  const singularForm = dateComponent.substr(0, dateComponent.length - 1);

  const thisDateNotation =
    singularForm === 'Day' ? 'Today' : `This ${singularForm}`;

  return thisDateNotation.toLowerCase();
};
