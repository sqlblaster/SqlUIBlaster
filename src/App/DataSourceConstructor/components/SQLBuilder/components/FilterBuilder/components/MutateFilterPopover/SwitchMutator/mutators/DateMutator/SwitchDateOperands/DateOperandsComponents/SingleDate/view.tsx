import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { DatePickerByOperandsType } from '../../DatePickerByOperandsType';
import { SingleDateViewProps } from './props';
import { singleDateStyles } from './styles';

// TODO: strech selects to the full available width
export const SingleDate = withStyles(singleDateStyles)((({
  operatorView,
  operands,
  classes,
  pickedVariables,
  handleVariablePicked,
  handleOperandsChange
}) => (
  <div className={classes.root}>
    {operatorView}
    <DatePickerByOperandsType
      operands={operands}
      onOperandsChange={handleOperandsChange}
      variable={pickedVariables}
      onVariablePicked={handleVariablePicked}
    />
  </div>
)) as React.FC<SingleDateViewProps>);
