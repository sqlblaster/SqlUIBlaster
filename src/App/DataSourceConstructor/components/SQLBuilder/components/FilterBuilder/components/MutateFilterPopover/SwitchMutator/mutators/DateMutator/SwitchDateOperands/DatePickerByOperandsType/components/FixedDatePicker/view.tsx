import { InputAdornment, Typography, withStyles } from '@material-ui/core';
import AccessTimeIcon from '@material-ui/icons/AccessTime';
import { DatePicker, TimePicker } from 'material-ui-pickers';
import * as React from 'react';
import { RemoveButton } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { FixedDatePickerViewProps } from './props';
import { fixedDatePickerStyles } from './styles';

export const addATimeText = 'Add a time';

export const FixedDatePicker = withStyles(fixedDatePickerStyles)(
  class extends React.Component<FixedDatePickerViewProps> {
    public render() {
      const {
        date,
        timeEnabled,
        limits,
        classes,
        handleDateChange,
        handleAddTimeClick,
        handleRemoveTimeClick
      } = this.props;

      return (
        <div className={classes.root}>
          <DatePicker
            value={date}
            onChange={handleDateChange}
            format='DD/MM/YYYY'
            margin='none'
            variant='outlined'
            style={{ width: '100%', marginTop: '10px' }}
            InputProps={{
              inputProps: {
                style: { padding: '7px 14px', fontSize: '13px' }
              }
            }}
            {...(limits ? limits : {})}
          />
          {!timeEnabled && (
            <label className={classes['add-time']} onClick={handleAddTimeClick}>
              <AccessTimeIcon
                style={{
                  marginRight: '5px',
                  fontSize: '16px',
                  fontWeight: 'bold'
                }}
              />
              <Typography color='inherit'>{addATimeText}</Typography>
            </label>
          )}
          {timeEnabled && (
            <TimePicker
              value={date}
              onChange={handleDateChange}
              margin='none'
              variant='outlined'
              style={{ width: '100%', marginTop: '10px' }}
              InputProps={{
                inputProps: {
                  style: { padding: '7px 14px', fontSize: '13px' }
                },
                endAdornment: (
                  <InputAdornment position='end'>
                    <RemoveButton onClick={handleRemoveTimeClick} />
                  </InputAdornment>
                )
              }}
            />
          )}
        </div>
      );
    }
  }
);
