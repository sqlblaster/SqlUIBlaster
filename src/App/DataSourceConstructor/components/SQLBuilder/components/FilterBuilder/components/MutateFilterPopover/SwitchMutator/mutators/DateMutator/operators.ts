enum DateOperatorEnum {
  'Previous',
  'Next',
  'Before',
  'After',
  'On',
  'Between',
  'Is empty',
  'Is not empty'
}

export type DateOperator = keyof typeof DateOperatorEnum;

export const dateOperators: DateOperator[] = [
  'Previous',
  'Next',
  'Before',
  'After',
  'On',
  'Between',
  'Is empty',
  'Is not empty'
];
