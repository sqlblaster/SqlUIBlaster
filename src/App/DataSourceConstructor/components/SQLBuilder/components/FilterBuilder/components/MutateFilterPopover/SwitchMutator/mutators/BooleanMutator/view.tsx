import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { CheckedSelect } from '../../CheckedSelect';
import { BooleanOperator, booleanOperators } from './operators';
import { BooleanMutatorViewProps } from './props';
import { booleanMutatorStyles } from './styles';

export const BooleanMutator = withStyles(booleanMutatorStyles)(
  class extends React.Component<BooleanMutatorViewProps> {
    public render() {
      const {
        operator,
        handleOperatorSelect,
        titleComponent,
        classes,
        submitMutationComponent
      } = this.props;

      return (
        <div className={classes.root}>
          <div
            className={ `${classes['title-and-operator']} ${
              classes['common-padding']
            }` }
          >
            <div className={classes.title}>{titleComponent}</div>
            <div className={classes.operator}>
              <CheckedSelect<BooleanOperator>
                selectedItem={operator}
                items={booleanOperators}
                onSelect={handleOperatorSelect}
              />
            </div>
          </div>
          <hr className={classes.hr} />
          <div
            className={ `${classes['mutator-action']} ${
              classes['common-padding']
            }` }
          >
            {submitMutationComponent}
          </div>
        </div>
      );
    }
  }
);
