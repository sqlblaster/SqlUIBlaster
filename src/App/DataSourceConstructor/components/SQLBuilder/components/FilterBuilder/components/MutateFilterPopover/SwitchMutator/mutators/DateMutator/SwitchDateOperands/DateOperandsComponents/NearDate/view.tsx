import { Checkbox, FormControlLabel, TextField, withStyles } from '@material-ui/core';
import * as React from 'react';
import { CheckedSelect } from '../../../../../CheckedSelect';
import { PluralDateComponent, pluralDateComponents } from './plural-date-components';
import { NearDateViewProps } from './props';
import { nearDateStyles } from './styles';
import { getThisDateComponentNotation } from './utils';

export const NearDate = withStyles(nearDateStyles)((({
  operands: { shiftAmount, dateComponentType, includeCurrentDate },
  operatorView,
  classes,
  renderBottomSection,
  handleShiftAmountChange,
  handleDateComponentTypeChange,
  handleIncludeCurrentDateChange
}) => {
  return (
    <>
      <div className={classes['near-date']}>
        {operatorView}
        <TextField
          className={classes['text-field']}
          margin='dense'
          variant='outlined'
          value={shiftAmount}
          onChange={handleShiftAmountChange}
          inputProps={{ type: 'number', className: classes['shift-amount'] }}
        />
        <CheckedSelect<PluralDateComponent>
          selectedItem={dateComponentType}
          items={pluralDateComponents}
          onSelect={handleDateComponentTypeChange}
        />
      </div>
      {renderBottomSection(
        <FormControlLabel
          className={classes.FormControlLabel}
          control={
            <Checkbox
              id={'include-date-checkbox'}
              checked={includeCurrentDate}
              onChange={handleIncludeCurrentDateChange}
              color='primary'
              style={{ padding: '0px' }}
            />
          }
          htmlFor={'include-date-checkbox'}
          label={
            <div className={classes.label}>
              Include
              <span className={classes['date-component']}>
                {` ${getThisDateComponentNotation(dateComponentType)}`}
              </span>
            </div>
          }
        />
      )}
    </>
  );
}) as React.FC<NearDateViewProps>);
