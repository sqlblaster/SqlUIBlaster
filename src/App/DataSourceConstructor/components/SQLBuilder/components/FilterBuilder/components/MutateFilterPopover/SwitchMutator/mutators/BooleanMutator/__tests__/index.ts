import {
  fireEvent,
  getAllByDisplayValue as getAllByDisplayValueOn,
  getByDisplayValue as getByDisplayValueOn
} from 'react-testing-library';
import {
  getGeneratedQuery,
  render
} from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/__tests__/render.utils';
import {
  clickAddTheFilter,
  clickUpdateTheFilter
} from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/__tests__/utils/filter-utils';
import { selectATable } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/TableSelector/__tests__/utils/select-table';
import {
  getByText$,
  regex
} from 'src/components/query-builder/App/DataSourceConstructor/test-utils/utils';
import { clickAddAFilter } from '../../../../../AddFilter/__tests__/utils/click-add-a-filter';
import { mutatorPanelId } from '../../../../view';
import { BooleanOperator, booleanOperators } from '../operators';

describe(`boolean operations`, () => {
  it(`should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  `, async () => {
    const { getByText, getByTestId, debug, componentNode } = render();

    const columns = selectATable();
    const columnName = columns[4].name;

    const clickOnFilter = () => {
      fireEvent.click(getByText(regex(columnName)));
    };

    const checkFilterDescription = (operator: BooleanOperator) => {
      expect(
        getByText$(componentNode, new RegExp(`^${columnName}.+$`, 'i'))
          .textContent
      ).toMatchSnapshot(
        `filter description for '${operator}' boolean operator`
      );
    };

    const checkOperation = (
      nextOperator: BooleanOperator,
      prevOperator?: BooleanOperator
    ) => {
      clickOnFilter();
      prevOperator &&
        fireEvent.click(getByText(new RegExp(`^${prevOperator}.*$`)));
      prevOperator &&
        fireEvent.click(getByText(new RegExp(`^.*${nextOperator}$`)));
      clickUpdateTheFilter();

      checkFilterDescription(nextOperator);

      expect(getGeneratedQuery()).toMatchSnapshot(
        `'${nextOperator}' boolean operator`
      );
    };

    // default boolean filter
    clickAddAFilter();
    fireEvent.click(getByText(regex(columnName)));
    clickAddTheFilter();

    checkFilterDescription('Is true');

    expect(getGeneratedQuery()).toMatchSnapshot(`default boolean operator`);

    // Check all operators
    for (let i = 1; i < booleanOperators.length; i++) {
      const nextOperator = booleanOperators[i];
      const prevOperator = booleanOperators[i - 1];
      checkOperation(nextOperator, prevOperator);
    }
  });
});
