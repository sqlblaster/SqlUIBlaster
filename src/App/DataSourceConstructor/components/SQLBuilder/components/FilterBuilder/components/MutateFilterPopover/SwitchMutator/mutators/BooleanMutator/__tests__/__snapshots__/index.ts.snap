// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : 'Is empty' boolean operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Paid\\" IS NULL
LIMIT 2000
"
`;

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : 'Is false' boolean operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Paid\\" = FALSE
LIMIT 2000
"
`;

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : 'Is not empty' boolean operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Paid\\" IS NOT NULL
LIMIT 2000
"
`;

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : default boolean operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Paid\\" = TRUE
LIMIT 2000
"
`;

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : filter description for 'Is empty' boolean operator 1`] = `"Paidis empty"`;

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : filter description for 'Is false' boolean operator 1`] = `"Paidis false"`;

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : filter description for 'Is not empty' boolean operator 1`] = `"Paidis not empty"`;

exports[`boolean operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-boolean operands,
      should allow non-integer operands
  : filter description for 'Is true' boolean operator 1`] = `"Paidis true"`;
